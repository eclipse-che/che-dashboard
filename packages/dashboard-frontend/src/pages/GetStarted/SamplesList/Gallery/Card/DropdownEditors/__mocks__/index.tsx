/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/GetStarted/SamplesList/Gallery/Card/DropdownEditors';

export class DropdownEditors extends React.PureComponent<Props> {
  render(): React.ReactNode {
    const { editors, onEditorSelect } = this.props;
    return (
      <div data-testid="card-actions-dropdown">
        <div>spanDropdownEditors</div>
        {editors.map(editor => (
          <button
            data-testid="card-action"
            key={editor.id}
            onClick={() => onEditorSelect(editor.id)}
          >
            {editor.name}
          </button>
        ))}
        <button onClick={() => onEditorSelect(editors[0].id)}>Select Editor</button>
      </div>
    );
  }
}
