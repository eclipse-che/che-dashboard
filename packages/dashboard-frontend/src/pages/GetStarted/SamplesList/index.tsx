/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Flex, FlexItem } from '@patternfly/react-core';
import { History } from 'history';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import SamplesListGallery from '@/pages/GetStarted/SamplesList/Gallery';
import { SamplesListHeader } from '@/pages/GetStarted/SamplesList/Header';
import SamplesListToolbar from '@/pages/GetStarted/SamplesList/Toolbar';
import { buildFactoryLocation, toHref } from '@/services/helpers/location';
import { che } from '@/services/models';
import { AppState } from '@/store';
import {
  DevfileRegistryMetadata,
  selectMetadataFiltered,
} from '@/store/DevfileRegistries/selectors';
import { selectPvcStrategy } from '@/store/ServerConfig/selectors';

export type Props = {
  history: History;
} & MappedProps;

type State = {
  isTemporary: boolean;
};

class SamplesList extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isTemporary: this.props.preferredStorageType === 'ephemeral' ? true : false,
    };
  }

  private handleTemporaryStorageChange(isTemporary: boolean): void {
    this.setState({ isTemporary });
  }

  private getStorageType(): che.WorkspaceStorageType {
    const { preferredStorageType } = this.props;
    const { isTemporary } = this.state;

    if (isTemporary) {
      return 'ephemeral';
    }

    return preferredStorageType === 'ephemeral' ? 'persistent' : preferredStorageType;
  }

  private async handleSampleCardClick(
    metadata: DevfileRegistryMetadata,
    editorId: string | undefined,
  ): Promise<void> {
    const factoryUrlParams = new URLSearchParams({ url: metadata.links.v2 });

    if (editorId !== undefined) {
      factoryUrlParams.append('che-editor', editorId);

      const prebuiltDevWorkspace = metadata.links.devWorkspaces?.[editorId];
      if (prebuiltDevWorkspace !== undefined) {
        factoryUrlParams.append('devWorkspace', prebuiltDevWorkspace);
      }
    }

    const storageType = this.getStorageType();
    factoryUrlParams.append('storageType', storageType);

    const factoryLocation = buildFactoryLocation();
    factoryLocation.search = factoryUrlParams.toString();

    const factoryLink = toHref(this.props.history, factoryLocation);
    window.open(factoryLink, '_blank');
  }

  public render(): React.ReactElement {
    const { metadataFiltered } = this.props;
    return (
      <>
        <Flex direction={{ default: 'column' }}>
          <FlexItem spacer={{ default: 'spacerLg' }}>
            <SamplesListHeader />
          </FlexItem>
          <FlexItem grow={{ default: 'grow' }} spacer={{ default: 'spacerLg' }}>
            <SamplesListToolbar
              isTemporary={this.state.isTemporary}
              onTemporaryStorageChange={isTemporary =>
                this.handleTemporaryStorageChange(isTemporary)
              }
            />
          </FlexItem>
        </Flex>
        <SamplesListGallery
          metadataFiltered={metadataFiltered}
          onCardClick={(metadata, editorId) => this.handleSampleCardClick(metadata, editorId)}
        />
      </>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  metadataFiltered: selectMetadataFiltered(state),
  preferredStorageType: selectPvcStrategy(state),
});

const connector = connect(mapStateToProps);
type MappedProps = ConnectedProps<typeof connector>;
export default connector(SamplesList);
