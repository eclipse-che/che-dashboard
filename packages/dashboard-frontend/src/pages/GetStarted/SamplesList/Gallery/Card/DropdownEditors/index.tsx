/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Dropdown,
  DropdownGroup,
  DropdownItem,
  DropdownPosition,
  KebabToggle,
} from '@patternfly/react-core';
import { CheckIcon } from '@patternfly/react-icons';
import React from 'react';

import TagLabel from '@/components/TagLabel';
import { PluginEditor } from '@/pages/GetStarted/SamplesList/Gallery';
import styles from '@/pages/GetStarted/SamplesList/Gallery/Card/DropdownEditors/index.module.css';

export type Props = {
  editors: PluginEditor[];
  onEditorSelect: (editorId: string) => void;
};

type State = {
  isExpanded: boolean;
};

export class DropdownEditors extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      isExpanded: false,
    };
  }

  private handleClick(editorId: string): void {
    this.props.onEditorSelect(editorId);
  }

  private toggleExpand(isExpanded: boolean): void {
    this.setState({ isExpanded });
  }

  private getDropdownItemGroup(): React.ReactNode[] {
    const { editors } = this.props;

    return [
      <DropdownGroup
        className={styles.dropdownEditorGroup}
        label="Choose an editor"
        key="editor-group"
      >
        {editors.map(editor => {
          return (
            <DropdownItem
              key={editor.id}
              tooltip={editor.displayName}
              onClick={() => this.handleClick(editor.id)}
              data-testid="card-action"
            >
              <span className={styles.editorTitle}>{editor.name}</span>
              <TagLabel version={editor.version} />
              {editor.isDefault && (
                <CheckIcon data-testid="checkmark" className={styles.checkIcon} />
              )}
            </DropdownItem>
          );
        })}
      </DropdownGroup>,
    ];
  }

  public render(): React.ReactElement {
    const { isExpanded } = this.state;

    const dropdownItems = this.getDropdownItemGroup();

    return (
      <Dropdown
        style={{ whiteSpace: 'nowrap' }}
        toggle={<KebabToggle onToggle={isExpanded => this.toggleExpand(isExpanded)} />}
        isOpen={isExpanded}
        position={DropdownPosition.right}
        dropdownItems={dropdownItems}
        isPlain
        data-testid="card-actions-dropdown"
        onClick={e => e.stopPropagation()}
      />
    );
  }
}
