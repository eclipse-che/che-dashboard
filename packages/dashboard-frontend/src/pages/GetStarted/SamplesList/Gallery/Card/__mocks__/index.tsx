/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/GetStarted/SamplesList/Gallery/Card';

export class SampleCard extends React.PureComponent<Props> {
  render(): React.ReactElement {
    const { metadata, onClick } = this.props;

    return (
      <div data-testid="sample-card">
        <div data-testid="sample-name">{metadata.displayName}</div>
        <button onClick={() => onClick()}>Select Sample</button>
      </div>
    );
  }
}
