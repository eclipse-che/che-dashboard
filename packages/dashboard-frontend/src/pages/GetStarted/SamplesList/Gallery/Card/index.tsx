/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Badge,
  Brand,
  Card,
  CardActions,
  CardBody,
  CardHeader,
  CardHeaderMain,
} from '@patternfly/react-core';
import { CubesIcon } from '@patternfly/react-icons';
import React from 'react';

import { PluginEditor, VISIBLE_TAGS } from '@/pages/GetStarted/SamplesList/Gallery';
import { DropdownEditors } from '@/pages/GetStarted/SamplesList/Gallery/Card/DropdownEditors';
import styles from '@/pages/GetStarted/SamplesList/Gallery/Card/index.module.css';
import { che } from '@/services/models';
import { convertIconToSrc } from '@/services/registry/devfiles';
import { DevfileRegistryMetadata } from '@/store/DevfileRegistries/selectors';

export type Props = {
  metadata: DevfileRegistryMetadata;
  editors: PluginEditor[];
  onClick: (editorId: string | undefined) => void;
};

export class SampleCard extends React.PureComponent<Props> {
  private getTags(): JSX.Element[] {
    const {
      metadata: { tags },
    } = this.props;

    const createTag = (text: string, key: number): React.ReactElement => {
      return (
        <Badge
          isRead
          style={{ whiteSpace: 'nowrap' }}
          key={`badge_${key}`}
          data-testid="card-badge"
        >
          {text.trim()}
        </Badge>
      );
    };

    return tags
      .filter(tag => VISIBLE_TAGS.indexOf(tag) !== -1)
      .map((item: string, index: number) => createTag(item, index));
  }

  private handleCardClick(editorId: string | undefined): void {
    this.props.onClick(editorId);
  }

  private buildIcon(metadata: che.DevfileMetaData): React.ReactElement {
    const props = {
      className: styles.sampleCardIcon,
      alt: metadata.displayName,
      'aria-label': metadata.displayName,
    };

    return metadata.icon ? (
      <Brand src={convertIconToSrc(metadata.icon)} {...props} data-testid="sample-card-icon" />
    ) : (
      <CubesIcon {...props} data-testid="default-card-icon" />
    );
  }

  render(): React.ReactElement {
    const { metadata, editors } = this.props;

    const tags = this.getTags();
    const devfileIcon = this.buildIcon(metadata);

    return (
      <Card
        isFlat
        isCompact
        isSelectable
        key={metadata.links.v2}
        onClick={() => this.handleCardClick(undefined)}
        className={'sample-card'}
        data-testid="sample-card"
      >
        <CardHeader
          toggleButtonProps={{
            id: 'toggle-button',
            'aria-label': 'toggle button',
            'aria-expanded': 'true',
          }}
          onClick={() => this.handleCardClick(undefined)}
        >
          <CardHeaderMain>{devfileIcon}</CardHeaderMain>
          <CardActions>
            {tags}
            <DropdownEditors
              onEditorSelect={editorId => this.handleCardClick(editorId)}
              editors={editors}
            />
          </CardActions>
        </CardHeader>
        <CardHeader>{metadata.displayName}</CardHeader>
        <CardBody>{metadata.description}</CardBody>
      </Card>
    );
  }
}
