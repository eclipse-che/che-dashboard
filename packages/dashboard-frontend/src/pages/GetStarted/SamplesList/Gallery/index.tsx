/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Button,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStatePrimary,
  EmptyStateVariant,
  Gallery,
  Title,
} from '@patternfly/react-core';
import { SearchIcon } from '@patternfly/react-icons';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { SampleCard } from '@/pages/GetStarted/SamplesList/Gallery/Card';
import {
  filterMostPrioritized,
  sortByPriority,
} from '@/pages/GetStarted/SamplesList/Gallery/filterEditors';
import { che } from '@/services/models';
import { AppState } from '@/store';
import * as DevfileRegistriesStore from '@/store/DevfileRegistries';
import { DevfileRegistryMetadata, EMPTY_WORKSPACE_TAG } from '@/store/DevfileRegistries/selectors';
import { selectEditors } from '@/store/Plugins/chePlugins/selectors';
import { selectDefaultEditor } from '@/store/Plugins/devWorkspacePlugins/selectors';

export type PluginEditor = che.Plugin & {
  isDefault: boolean;
};

export const VISIBLE_TAGS = ['Community', 'Tech-Preview', 'Devfile.io'];

const EXCLUDED_TARGET_EDITOR_NAMES = ['dirigible', 'jupyter', 'eclipseide', 'code-server'];

export type Props = MappedProps & {
  metadataFiltered: DevfileRegistryMetadata[];
  onCardClick: (metadata: DevfileRegistryMetadata, editorId: string | undefined) => void;
};

export class SamplesListGallery extends React.PureComponent<Props> {
  private handleCardClick(metadata: DevfileRegistryMetadata, editorId: string | undefined): void {
    this.props.onCardClick(metadata, editorId);
  }

  private prepareEditors(): PluginEditor[] {
    const { editors, defaultEditor } = this.props;

    const allowed = editors
      .filter(editor => !EXCLUDED_TARGET_EDITOR_NAMES.includes(editor.name))
      .map(editor => ({
        ...editor,
        isDefault: defaultEditor === editor.id,
      }));

    const sorted = sortByPriority(allowed);
    const filtered = filterMostPrioritized(sorted);

    return filtered;
  }

  private prepareMetadata(): DevfileRegistryMetadata[] {
    function sortByVisibleTag(a: DevfileRegistryMetadata, b: DevfileRegistryMetadata): -1 | 0 | 1 {
      const getVisibleTag = (metadata: DevfileRegistryMetadata) =>
        metadata.tags.filter(tag => VISIBLE_TAGS.includes(tag))[0];
      const tagA = getVisibleTag(a);
      const tagB = getVisibleTag(b);
      if (tagA === tagB) {
        return 0;
      }
      if (tagA === undefined || tagA < tagB) {
        return -1;
      }
      if (tagB === undefined || tagA > tagB) {
        return 1;
      }
      return 0;
    }

    function sortByEmptyWorkspaceTag(
      a: DevfileRegistryMetadata,
      b: DevfileRegistryMetadata,
    ): -1 | 0 | 1 {
      if (a.tags.includes(EMPTY_WORKSPACE_TAG) > b.tags.includes(EMPTY_WORKSPACE_TAG)) {
        return -1;
      }
      if (a.tags.includes(EMPTY_WORKSPACE_TAG) < b.tags.includes(EMPTY_WORKSPACE_TAG)) {
        return 1;
      }
      return 0;
    }

    function sortByDisplayName(a: DevfileRegistryMetadata, b: DevfileRegistryMetadata): -1 | 0 | 1 {
      if (a.displayName < b.displayName) {
        return -1;
      }
      if (a.displayName > b.displayName) {
        return 1;
      }
      return 0;
    }

    return this.props.metadataFiltered
      .sort(sortByDisplayName)
      .sort(sortByVisibleTag)
      .sort(sortByEmptyWorkspaceTag);
  }

  private buildCardsList(): React.ReactElement[] {
    const pluginEditors = this.prepareEditors();
    const metadata = this.prepareMetadata();

    return metadata.map(meta => (
      <SampleCard
        key={meta.links.v2}
        metadata={meta}
        editors={pluginEditors}
        onClick={(editorId: string | undefined) => this.handleCardClick(meta, editorId)}
      />
    ));
  }

  render(): React.ReactElement {
    const cards = this.buildCardsList();

    if (cards.length === 0) {
      return (
        <EmptyState variant={EmptyStateVariant.full}>
          <EmptyStateIcon icon={SearchIcon} />
          <Title headingLevel="h1">No results found</Title>
          <EmptyStateBody>
            No results match the filter criteria. Clear filter to show results.
          </EmptyStateBody>
          <EmptyStatePrimary>
            <Button variant="link" onClick={(): void => this.props.clearFilter()}>
              Clear filter
            </Button>
          </EmptyStatePrimary>
        </EmptyState>
      );
    }

    return <Gallery hasGutter={true}>{cards}</Gallery>;
  }
}

const mapStateToProps = (state: AppState) => ({
  editors: selectEditors(state),
  defaultEditor: selectDefaultEditor(state),
});

const connector = connect(mapStateToProps, {
  ...DevfileRegistriesStore.actionCreators,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(SamplesListGallery);
