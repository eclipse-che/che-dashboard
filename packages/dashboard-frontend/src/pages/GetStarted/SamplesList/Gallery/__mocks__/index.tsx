/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/GetStarted/SamplesList/Gallery';

export default class SamplesListGallery extends React.PureComponent<Props> {
  render() {
    const { onCardClick, metadataFiltered = [] } = this.props;
    return (
      <div data-testid="samples-list-gallery">
        <div>Samples List Gallery</div>
        {metadataFiltered.map(metadata => (
          <div key={metadata.links.v2} data-testid="sample-card">
            {metadata.displayName}

            <button
              onClick={() => {
                onCardClick(metadata);
              }}
            >
              Select Sample
            </button>
          </div>
        ))}
      </div>
    );
  }
}
