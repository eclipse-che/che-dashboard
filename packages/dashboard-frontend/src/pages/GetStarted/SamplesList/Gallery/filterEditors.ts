/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { PluginEditor } from '@/pages/GetStarted/SamplesList/Gallery';

/**
 * Filter the editors to leave the first occurrence of each editor name.
 */
export function filterMostPrioritized(editors: PluginEditor[]): PluginEditor[] {
  const editorNames = new Set();
  const filteredEditors = editors.filter(editor => {
    if (!editorNames.has(editor.name)) {
      editorNames.add(editor.name);
      return true;
    }
    return false;
  });

  return filteredEditors;
}

/**
 * Sort editors by name and version. The priority is defined as follows:
 * 1. Default editor
 * 2. Version insiders
 * 3. Version next
 * 4. Version latest
 * The rest of the editors are sorted by id.
 */
export function sortByPriority(editors: PluginEditor[]) {
  function sortFn(editorA: PluginEditor, editorB: PluginEditor) {
    const priority = ['isDefault', 'insiders', 'next', 'latest'];

    if (editorA.name === editorB.name) {
      for (const prop of priority) {
        if (editorA.version === prop) return -1;
        if (editorB.version === prop) return 1;
      }
    }
    return editorA.id.localeCompare(editorB.id);
  }

  return editors.sort(sortFn);
}
