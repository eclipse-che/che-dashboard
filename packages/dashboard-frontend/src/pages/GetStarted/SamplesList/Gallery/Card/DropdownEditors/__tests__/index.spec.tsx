/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Dropdown, KebabToggle } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { PluginEditor } from '@/pages/GetStarted/SamplesList/Gallery';
import { DropdownEditors } from '@/pages/GetStarted/SamplesList/Gallery/Card/DropdownEditors';
import getComponentRenderer, { screen, within } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const onItemClick = jest.fn();

describe('DropdownEditors component', () => {
  let editors: PluginEditor[];

  beforeEach(() => {
    editors = [
      {
        id: 'che-incubator/che-code/insiders',
        name: 'che-code',
        description:
          'Microsoft Visual Studio Code - Open Source IDE for Eclipse Che - Insiders build',
        version: 'insiders',
        isDefault: true,
      },
      {
        id: 'che-incubator/che-idea/next',
        name: 'che-idea',
        description: 'JetBrains IntelliJ IDEA Community IDE for Eclipse Che - next',
        version: 'next',
        isDefault: false,
      },
    ] as PluginEditor[];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot(editors);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should have editor items', async () => {
    renderComponent(editors);

    const dropdown = screen.getByTestId('card-actions-dropdown');

    const menuButton = within(dropdown).queryByRole('button');
    expect(menuButton).not.toBeNull();

    userEvent.click(menuButton as HTMLElement);

    const menuitems = await screen.findAllByRole('menuitem');

    expect(menuitems.length).toEqual(2);

    const codeItem = menuitems[0];
    expect(codeItem.textContent).toContain('che-code');

    const codeItemCheckMark = within(codeItem).queryByTestId('checkmark');
    expect(codeItemCheckMark).not.toBeNull();

    const ideaItem = menuitems[1];
    expect(ideaItem.textContent).toContain('che-idea');

    const ideaItemCheckMark = within(ideaItem).queryByTestId('checkmark');
    expect(ideaItemCheckMark).toBeNull();
  });

  it('should handle "onClick" events', async () => {
    renderComponent(editors);

    const dropdown = screen.getByTestId('card-actions-dropdown');

    const menuButton = within(dropdown).queryByRole('button');
    expect(menuButton).not.toBeNull();

    userEvent.click(menuButton as HTMLElement);

    const menuitems = await screen.findAllByRole('menuitem');

    const codeItem = menuitems[0];
    userEvent.click(codeItem);

    expect(onItemClick).toHaveBeenCalledWith('che-incubator/che-code/insiders');
  });
});

function getComponent(editors: PluginEditor[]): React.ReactElement {
  return (
    <Dropdown
      isOpen={true}
      toggle={<KebabToggle />}
      dropdownItems={[
        <DropdownEditors key="test" onEditorSelect={onItemClick} editors={editors} />,
      ]}
    />
  );
}
