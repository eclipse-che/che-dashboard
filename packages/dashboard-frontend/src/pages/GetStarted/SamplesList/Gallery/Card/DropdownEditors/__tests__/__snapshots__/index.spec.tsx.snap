// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownEditors component snapshot 1`] = `
<div
  className="pf-c-dropdown pf-m-expanded"
  data-ouia-component-id="OUIA-Generated-Dropdown-1"
  data-ouia-component-type="PF4/Dropdown"
  data-ouia-safe={true}
>
  <button
    aria-expanded={true}
    aria-haspopup={true}
    aria-label="Actions"
    className="pf-c-dropdown__toggle"
    disabled={false}
    id="pf-dropdown-toggle-id-0"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      aria-labelledby={null}
      fill="currentColor"
      height="1em"
      role="img"
      style={
        {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 192 512"
      width="1em"
    >
      <path
        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
      />
    </svg>
  </button>
  <ul
    aria-labelledby="pf-dropdown-toggle-id-0"
    autoFocus={false}
    className="pf-c-dropdown__menu"
    hidden={false}
    role="menu"
  >
    <div
      className="pf-c-dropdown pf-m-align-right"
      data-ouia-component-id="OUIA-Generated-Dropdown-2"
      data-ouia-component-type="PF4/Dropdown"
      data-ouia-safe={true}
      data-testid="card-actions-dropdown"
      onClick={[Function]}
      style={
        {
          "whiteSpace": "nowrap",
        }
      }
    >
      <button
        aria-expanded={false}
        aria-haspopup={true}
        aria-label="Actions"
        className="pf-c-dropdown__toggle pf-m-plain"
        disabled={false}
        id="pf-dropdown-toggle-id-1"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          fill="currentColor"
          height="1em"
          role="img"
          style={
            {
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 192 512"
          width="1em"
        >
          <path
            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
          />
        </svg>
      </button>
    </div>
  </ul>
</div>
`;
